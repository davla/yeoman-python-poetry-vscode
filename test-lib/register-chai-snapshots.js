import chai from "chai";
import { jestSnapshotPlugin as chaiSnapshot } from "mocha-chai-jest-snapshot";

chai.use(chaiSnapshot());
