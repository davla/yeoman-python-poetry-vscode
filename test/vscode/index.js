import fs from "node:fs/promises";
import path from "node:path";

import yeomanTest from "yeoman-test";

import "../../test-lib/register-chai-snapshots.js";
import VSCodeGenerator from "../../generators/vscode/index.js";
import {
  readJsonInCwd,
  readTomlInCwd,
  writeTomlInCwd,
} from "../../test-lib/file-system.js";
import restoreRunResult from "../../test-lib/generator-hooks.js";
import {
  cleanupSystemAccessStubs,
  setupSystemAccessStubs,
} from "../../test-lib/system-access-stubs.js";

async function writeVsCodeConfig(fileName, content, dstDir) {
  const filePath = path.join(dstDir, ".vscode", fileName);
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, JSON.stringify(content));
}

const toPosixPath = (p) => p.replace(path.sep, path.posix.sep);

describe("python-poetry-vscode:vscode", () => {
  beforeEach(function () {
    this.stubs = setupSystemAccessStubs();
    this.generator = yeomanTest.run(VSCodeGenerator);
  });

  afterEach(cleanupSystemAccessStubs);

  describe("vscode config", () => {
    afterEach(restoreRunResult);

    for (const fileName of ["settings", "extensions"]) {
      const filePath = path.join(".vscode", `${fileName}.json`);

      it(`creates ${toPosixPath(filePath)} in JSON format`, async function () {
        this.runResult = await this.generator;
        this.runResult.assertFile(filePath);
        await readJsonInCwd(this.runResult, filePath).should.be.fulfilled;
      });

      it(`populates ${toPosixPath(filePath)}`, async function () {
        this.runResult = await this.generator;
        (await readJsonInCwd(this.runResult, filePath)).should.matchSnapshot();
      });
    }

    it(`merges .vscode/settings.json with existing content`, async function () {
      const content = {
        "python.linting.ignorePatterns": [".venv"],
        "[python]": {
          "editor.guides.indentation": true,
        },
      };
      this.runResult = await this.generator.doInDir(
        writeVsCodeConfig.bind(this.generator, "settings.json", content)
      );
      const fileInDst = path.join(".vscode", "settings.json");
      (await readJsonInCwd(this.runResult, fileInDst)).should.containSubset({
        "python.linting.ignorePatterns": ["**/site-packages/**/*.py", ".venv"],
        "[python]": {
          "editor.defaultFormatter": "ms-python.black-formatter",
          "editor.guides.indentation": true,
        },
      });
    });

    it(`merges .vscode/extensions.json with existing content`, async function () {
      const existingContent = {
        recommendations: ["BriteSnow.vscode-toggle-quotes"],
      };
      this.runResult = await this.generator.doInDir(
        writeVsCodeConfig.bind(
          this.generator,
          "extensions.json",
          existingContent
        )
      );
      const fileInDst = path.join(".vscode", "extensions.json");
      (await readJsonInCwd(this.runResult, fileInDst)).should.containSubset({
        recommendations: [
          "ms-python.black-formatter",
          "BriteSnow.vscode-toggle-quotes",
        ],
      });
    });
  });

  describe("poetry config", () => {
    afterEach(restoreRunResult);

    for (const fileName of ["poetry", "pyproject"]) {
      const filePath = fileName + ".toml";

      it(`creates ${toPosixPath(filePath)} in TOML format`, async function () {
        this.runResult = await this.generator;
        this.runResult.assertFile(filePath);
        await readTomlInCwd(this.runResult, filePath).should.be.fulfilled;
      });

      it(`populates ${toPosixPath(filePath)}`, async function () {
        this.runResult = await this.generator;
        (await readTomlInCwd(this.runResult, filePath)).should.matchSnapshot();
      });
    }

    it("merges poetry.toml with existing content", async function () {
      const existingContent = {
        virtualenvs: { create: true },
        installer: { parallel: false },
      };
      this.runResult = await this.generator.doInDir((dstDir) =>
        writeTomlInCwd(dstDir, "poetry.toml", existingContent)
      );
      (await readTomlInCwd(this.runResult, "poetry.toml")).should.containSubset(
        {
          virtualenvs: { create: true, "in-project": true },
          installer: { parallel: false },
        }
      );
    });

    it("merges pyproject.toml with existing content", async function () {
      const existingContent = {
        tool: {
          poetry: {
            dependencies: { pylint: "^2.15.0" },
            name: "existing_package",
          },
        },
      };
      this.runResult = await this.generator.doInDir((dstDir) =>
        writeTomlInCwd(dstDir, "pyproject.toml", existingContent)
      );
      (
        await readTomlInCwd(this.runResult, "pyproject.toml")
      ).should.containSubset({
        tool: {
          poetry: {
            dependencies: {
              black: "^22.12.0",
              isort: "^5.12.0",
              pylint: "^2.15.0",
            },
            name: "existing_package",
          },
        },
      });
    });
  });

  describe("install", () => {
    afterEach(restoreRunResult);

    it("doesn't run poetry install", async function () {
      this.runResult = await this.generator;
      this.stubs.spawnCommand.should.not.have.been.calledWith("poetry", [
        "install",
      ]);
    });
  });
});
