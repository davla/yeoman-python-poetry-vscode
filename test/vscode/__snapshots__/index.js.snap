// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`python-poetry-vscode:vscode poetry config populates poetry.toml 1`] = `
Object {
  "virtualenvs": Object {
    "in-project": true,
  },
}
`;

exports[`python-poetry-vscode:vscode poetry config populates pyproject.toml 1`] = `
Object {
  "tool": Object {
    "poetry": Object {
      "dependencies": Object {
        "black": "^22.12.0",
        "isort": "^5.12.0",
      },
      "group": Object {
        "dev": Object {
          "dependencies": Object {
            "pytest": "^7.3.1",
          },
        },
      },
    },
  },
}
`;

exports[`python-poetry-vscode:vscode vscode config populates .vscode/extensions.json 1`] = `
Object {
  "recommendations": Array [
    "bungcip.better-toml",
    "editorconfig.editorconfig",
    "ms-python.black-formatter",
    "ms-python.isort",
    "ms-python.python",
  ],
  "unwantedRecommendations": Array [],
}
`;

exports[`python-poetry-vscode:vscode vscode config populates .vscode/settings.json 1`] = `
Object {
  "[python]": Object {
    "editor.codeActionsOnSave": Object {
      "source.organizeImports": true,
    },
    "editor.defaultFormatter": "ms-python.black-formatter",
    "editor.formatOnSave": true,
    "editor.rulers": Array [
      88,
    ],
  },
  "black-formatter.importStrategy": "fromEnvironment",
  "isort.args": Array [
    "--profile",
    "black",
  ],
  "isort.importStrategy": "fromEnvironment",
  "python.formatting.provider": "none",
  "python.linting.ignorePatterns": Array [
    "**/site-packages/**/*.py",
  ],
  "python.testing.pytestEnabled": true,
  "python.testing.unittestEnabled": false,
}
`;
